rust_src/src/filelock.rs

```diff
@@ -2,13 +2,277 @@
+                .read_to_string(&mut content)?;
+            Ok(Some(content))
+        }
+        None => Ok(None),
+    }
+}
+
+/// Reads the target of the symbolic link at `path` and converts it
+/// to a [`String`]. Fails if the target is not valid UTF-8.
+fn read_link_as_string(path: &Path) -> Result<String> {
+    path.read_link()?
+        .into_os_string()
+        .into_string()
+        .map_err(|_| {
+            Error::new(
+                InvalidData,
+                format!("Target of link '{}' is not valid UTF-8", path.display()),
+            )
+        })
+}
+
+/// Reads the info string for the lock file `path`.
+fn read_lock_data(path: &Path) -> Result<String> {
+    loop {
+        match read_link_as_string(path) {
+            ok @ Ok(_) => return ok,
+
+            Err(ref e) if e.kind() == InvalidInput => match read_nofollow(path, MAX_LOCK_INFO)? {
+                Some(target) => return Ok(target),
+                None => (),
+            },
+
+            err => return err,
+        }
+
+        // `read_link` saw a non-symlink, but `open` saw a symlink.
+        // The former must have been removed and replaced by the latter.
+        // Try again.
+        unsafe { maybe_quit() }
+    }
+}
+
+/// Attempts to read and parse the lock information from the lock file `path`.
+/// Returns [`None`] if the lock file does not exist.
+fn read_lock_info(path: &Path) -> Result<Option<LockInfo>> {
+    match read_lock_data(path) {
+        Ok(data) => {
+            let info = LockInfo::parse(&data).ok_or(Error::new(
+                InvalidData,
+                format!("Invalid lock information in '{}'", path.display()),
+            ))?;
+            Ok(Some(info))
+        }
+
+        Err(ref e) if e.kind() == NotFound => Ok(None),
+
+        Err(e) => Err(e),
+    }
+}
+
+/// Returns [`true`] if a process with PID `pid` is running.
+#[cfg(unix)]
+fn process_exists(pid: i32) -> bool {
+    match unsafe { libc::kill(pid, 0) } {
+        0 => true,
+        // `PermissionDenied` indicates that there _is_ a process but it’s
+        // owned by another user
+        _ => Error::last_os_error().kind() == PermissionDenied,
+    }
+}
+
+/// Returns the system’s last boot time as Unix epoch seconds.
+fn get_boot_time() -> Result<i64> {
+    use systemstat::Platform;
+    let sys = systemstat::System::new();
+    sys.boot_time().map(|t| t.timestamp())
+}
+
+/// Returns [`true`] if there is no boot time in the lock info or if the
+/// boot time is within one second of the system’s boot time.
+fn boot_time_within_one_second(info: &LockInfo) -> bool {
+    info.boot_time.map_or(true, |boot_time| {
+        let system_boot_time = get_boot_time().unwrap_or(0);
+        (boot_time - system_boot_time).abs() <= 1
+    })
+}
+
+/// Returns the current state of the lock from lock file `path`.
+fn current_lock_owner(path: &Path) -> Result<LockState> {
+    use LockState::*;
+
+    match read_lock_info(path)? {
```

Comment:
This is another example where perhaps saving the result and returning `Ok(result)` is easier to follow.